text/0477-policy-validation.md

```diff
@@ -0,0 +1,553 @@
+
+```json
+"validations": {
+    "package": "cdk-cfn-guard",
+    "class": "CfnFuardValidation",
+    "version": "1.2.3"
+}
+```
+
+The CLI would read this file, load the appropriate plugins either from the filesystem or directly from NPM, and call
+them to get a validation report.
+
+We eventually discarded this option for the following reasons:
+
+* All plugins would have to be written in TypeScript/JavaScript for them to be consumable by the CLI.
+* The CLI lacks contextual information that would be useful for tracing the problem back to where it was defined.
+* If an application is synthesized without using the CDK CLI, it’s not subject to policy validation.
+
+#### CloudFormation hooks integration
+
+Instead of integrating with each policy tool individually, we have considered the option of developing a single
+mechanism, that would integrate with CloudFormation hooks. The idea was to invoke the Lambda function underlying a
+CloudFormation hook and interpret its results in a similar way CloudFormation, that is, by failing synthesis if the
+function reported a validation failure.
+
+However, the Lambda functions invoked by CloudFormation hooks are not deployed to the customer’s account, but to an
+account owned by AWS. This makes this solution not feasible.
+
+### What are the drawbacks of this solution?
+
+In many cases, plugins will need a third party tool installed on the developer's computer, such as a CLI. These tools
+are installed and managed separately from the CDK. This isn't a drawback — in fact, this separation of concerns is
+necessary for this feature to be extensible — but it places an extra burden on plugin authors. A developer of a CDK
+application that uses a given plugin may not have the necessary tools installed beforehand, when they synthesize the app
+for the first time. Plugin authors must then make sure that developers can install these tools with as little friction
+as possible to deliver a good user experience.
+
+### What is the high-level project plan?
+
+The first release of this feature will include integration with CloudFormation Guard. To make sure that our design is
+not tied to a specific vendor, we will also work on a prototype for the integration with a second policy checking tool (
+TBD). As usual with new CDK features, policy checking at synth time will be marked as experimental while we collect
+feedback from the community, fix bugs and stabilize the API. Once it’s promoted to stable, we’ll go back to the
+integration prototype and release it as a second officially supported integration.
+
+### Are there any open issues that need to be addressed later?
+
+No.
+
+## Appendix
+
+### Appendix A - High level design
+
+![Class diagram for policy validation](https://github.com/aws/aws-cdk-rfcs/blob/otaviom/rfc-0477/images/policy-validation/validation.png?raw=true)
+
+The main abstraction of the model is that of a validation plugin. A plugin encapsulates a piece of logic that includes
```

Comment:
Who installs the plugin itself? The framework or the user?

Reply:
Yeap

Reply:
That should be the responsibility of the plugin.